source/model/backtraceMetrics.ts

```diff
@@ -0,0 +1,228 @@
+
+  private readonly timestamp = currentTimestamp()
+  private readonly userAgent = USER_AGENT;
+  private readonly applicationName = APP_NAME;
+  private readonly applicationVersion = VERSION;
+
+  private readonly guid = getBacktraceGUID();
+
+  public hostname = 'https://events.backtrace.io';
+
+  constructor(configuration: BacktraceClientOptions) {
+    if (!configuration.endpoint) {
+      throw new Error(`Backtrace: missing 'endpoint' option.`);
+    }
+    const {universe, token} = getEndpointParams(configuration.endpoint)
+
+    if(!universe) {
+      throw new Error(`Backtrace: 'universe' could not be parsed from the endpoint.`);
+    }
+
+    if(!token && !configuration.token) {
+      throw new Error(`Backtrace: missing 'token' option or it could not be parsed from the endpoint.`);
+    }
+
+    this.universe = universe;
+    this.token = configuration.token || token || ''; // default '' case needed since type checking isn't certain that type can not be undefined.
+    this.timeout = configuration.timeout;  
+
+    this.persistSession(); // Create/persist session on construction.
+    // Persist session if page is focused on heartbeat interval
+    setInterval(() => this.persistIfFocused(), this.heartbeatInterval);
+  }
+
+  /**
+   * Handle persisting of session. When called, will create or manage current session.
+   * when appropriate.
+   */
+  private persistSession(): void {
+    const sessionId = this.getSessionId() || uuid();
+
+    // If sessionId is not set, create new session. Send unique and app launch events.
+    if (!sessionId) {
+      const newSessionId = this.createNewSession();
+      this.sendUniqueEvent(this.guid, newSessionId);
+      // An "application launch" loosely / temporarily means first session creation.
+      this.sendSummedEvent(this.guid, newSessionId, 'Application Launches');
+
+      // If sessionId is set and lastActive is over persistenceInterval, create new session and send unique event.
+    } else if (this.getLastActive() < (this.timestamp * SEC_TO_MILLIS) - this.persistenceInterval) {
+      this.createNewSession();
+      this.sendUniqueEvent(this.guid, sessionId);
+    }
+    // Can add else clause to send interval summed events.
+    // ex. "session length" by sending lastActive - sessionStart.
+    // This can get a sense of "session length" / "time spent on app", "error free minutes", "avg time per error" etc.
+
+    this.setLastActive(this.timestamp); // update lastActive. If this function is called, user is active.
+  }
+
+  /**
+   * Persist session if page is focused.
+   */
+  private persistIfFocused() : void {
+    if(!document.hidden) {
+      this.persistSession()
+    }
+  }
+
+  /**
+   * Send POST to unique-events API endpoint
+   */
+  private async sendUniqueEvent(guid: string, sessionId: string): Promise<void> {
+    const endpoint = `${this.hostname}/api/unique-events/submit?universe=${this.universe}&token=${this.token}`;
```

Comment:
do we need to create the endpoint url every time when we make a request. Not obvious thing, but every allocation require garbage collector to clean up stuff. It's a URL but it will make a difference in game engines/WebGL apps or UI apps that push web browsers to the limit

Reply:
hostname is now only configurable/changed construction, so URL is now a constant and also created once in construction.